opennav_following/src/following_server.cpp

```diff
@@ -0,0 +1,604 @@
+    result->error_msg = std::string("Transform error: ") + e.what();
+    RCLCPP_ERROR(get_logger(), result->error_msg.c_str());
+    result->error_code = FollowObject::Result::TF_ERROR;
+  } catch (opennav_following::FailedToDetectObject & e) {
+    result->error_msg = e.what();
+    RCLCPP_ERROR(get_logger(), result->error_msg.c_str());
+    result->error_code = FollowObject::Result::FAILED_TO_DETECT_OBJECT;
+  } catch (opennav_following::FailedToControl & e) {
+    result->error_msg = e.what();
+    RCLCPP_ERROR(get_logger(), result->error_msg.c_str());
+    result->error_code = FollowObject::Result::FAILED_TO_CONTROL;
+  } catch (opennav_following::FollowingException & e) {
+    result->error_msg = e.what();
+    RCLCPP_ERROR(get_logger(), result->error_msg.c_str());
+    result->error_code = FollowObject::Result::UNKNOWN;
+  } catch (std::exception & e) {
+    result->error_msg = e.what();
+    RCLCPP_ERROR(get_logger(), result->error_msg.c_str());
+    result->error_code = FollowObject::Result::UNKNOWN;
+  }
+
+  // Stop the robot and report
+  result->num_retries = num_retries_;
+  publishZeroVelocity();
+  following_action_server_->terminate_current(result);
+}
+
+bool FollowingServer::approachObject(
+  geometry_msgs::msg::PoseStamped & object_pose, const std::string & target_frame)
+{
+  rclcpp::Rate loop_rate(controller_frequency_);
+  while (rclcpp::ok()) {
+    publishFollowingFeedback(FollowObject::Feedback::CONTROLLING);
+
+    // Stop if cancelled/preempted
+    if (checkAndWarnIfCancelled<FollowObject>(following_action_server_, "follow_object") ||
+      checkAndWarnIfPreempted<FollowObject>(following_action_server_, "follow_object"))
+    {
+      return false;
+    }
+
+    // If we have a target frame_id from the goal, use that instead of pose detection
+    if (!target_frame.empty()) {
+      if (!getFramePose(target_frame, object_pose)) {
+        throw opennav_following::FailedToDetectObject(
+          "Failed to get pose in target frame: " + target_frame);
+      }
+    } else {
+      // Otherwise, use the traditional pose detection from topic
+      if (!getRefinedPose(object_pose)) {
+        throw opennav_following::FailedToDetectObject("Failed object detection");
+      }
+    }
+
+    // Get the pose at the distance we want to maintain from the object
+    // and transform the target_pose into base_frame
+    auto target_pose = getPoseAtDistance(object_pose, desired_distance_);
+    target_pose.header.stamp = rclcpp::Time(0);
+
+    // Stop and report success if goal is reached
+    if (isGoalReached(target_pose)) {
+      return true;
+    }
+
+    // The control law can get jittery when close to the end when atan2's can explode.
+    // Thus, we backward project the controller's target pose a little bit after the
+    // dock so that the robot never gets to the end of the spiral before its in contact
+    // with the dock to stop the docking procedure.
+    const double backward_projection = 0.25;
```

Comment:
If using the odom frame as the fixed frame, does this make sense to always be positive in value? The location of the robot w.r.t. the stable frame will make this dependent if this is in front, beyond, or even on the side of the obstacle rather than always being behind it.

I think what we need to do is find the vector from the detected obstacle, back out the `desired_distance_` from that, and then add back in the little bit along that vector. In essance, I think we really just want to do `desired_distance_ - 0.25` then can remove this block -- no?

Reply:
@ajtudela what is your resolution to this?